Refactor label painting by moving code from mouse bindings onto layer, adding staged history and paint event

DragaDoncila opened the pull request:
# Description
After some conversations on #4663 and with @jni and @tlambert03, I'm opening this draft PR to refactor some of the painting and history saving for the labels layer, also adding a paint event but this time in a more appropriate spot, on the layer itself.

The `labels_mouse_bindings` `draw` function was doing a *lot* of work checking layer attributes and computing coordinates, and this work really belongs on the layer itself. This PR moves some of that code into a private `_draw` method on the layer.

The `draw` function was also setting `layer._block_saving` as a way to let the layer know whether it should save one history atom, or start a new item. This led to some issues e.g. having to have a workaround for when the user started painting outside the canvas, and made it impossible to know, in the layer, when a larger mouse drag had finished. This PR therefore uses a `_staged_history` list to store history atoms, and adds a new `_finish_painting` method, whose job is to save staged history to real history, and emit the `paint` event with the history item. 

It may seem like we are now injecting "gui" logic into the layer, but I would say that the original `_block_saving` attribute was doing the same thing, but in a decentralised way. This PR at least makes sure that layer computations are not done in the mouse bindings, and that the layer has access to all the information it needs to emit its event. The `_finish_painting` method could also be used to support batch calls to `paint`, which otherwise would've needed toggling of the much more esoteric `_block_saving` attribute. We can also make `_finish_painting` a public method if we want to really support this.

This doesn't address @jni's concern on #4663 about emitting an event when someone splits labels or otherwise programmatically calls `paint`. I think we could address this by emitting in `paint`, but blocking the emission in `labels_mouse_bindings` `draw` until we call `_finish_painting`? Alternatively we could have two separate events e.g. `paint` and `painted` (or something more informative), but that may lead to performance issues - a concern originally raised in issue #4194.

## Type of change
<!-- Please delete options that are not relevant. -->
- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References
<!-- What resources, documentation, and guides were used in the creation of this PR? -->
<!-- If this is a bug-fix or otherwise resolves an issue, reference it here with "closes #(issue)" -->

# How has this been tested?
<!-- Please describe the tests that you ran to verify your changes. -->
- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

Commit:
Move label coordination and paint/fill calls to labels

Commit:
Move painting code into labels layer, define start_painting and end_painting, use staged history

Commit:
Docstrings and remove unnecessary start_painting

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
Remove unused import

jni commented on napari/layers/labels/_labels_mouse_bindings.py:
```diff
@@ -35,48 +29,20 @@ def draw(layer, event):
-    undo_item = layer._undo_history[-1]
-    if len(undo_item) == 1 and len(undo_item[0][0][0]) == 0:
-        layer._undo_history.pop()
+    layer._finish_painting()
```
@DragaDoncila so the thing that I always wanted to do here but never quite managed was something like:

```python
with layer.block_history():
    <painting_loop>
```

Then in the layer you would have:

```python
@contextmanager
def block_history(self):
    self._block_history = True
    yield
    self._block_history = False
```

Finally, *all* of paint, fill etc. would do:

```python
self.save_history(history_atom)
```

Finally, `self.save_history` would be the one function that would check `self._block_history`, and either append to the staging item or append the staging item to the history.

The advantages of this approach are:
- Use a context manager for something that smells very much like a context
- rather than move the append/add to history one level down to the paint and fill methods, move it all the way where it belongs in the save_history method

What do you think?

Czaki replied:
@jni 
when using `contextmanager` are used an exception is raised inside `with` clause in `yield` place. As your example does not contain a handling exception then stat could be blocked. 

See that in the documentation there is the suggestion to use `try: ... finally:`
https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager

It should look like this:

```python
@contextmanager
def block_history(self):
    prev = self._block_history
    self._block_history = True
    try:
        yield
    finally: 
        self._block_history = prev
```

jni replied:
Thanks for that snippet @Czaki! 🙏

Commit:
use context manager for saving to history or staged history

Commit:
precommit post commit

DragaDoncila commented on napari/layers/labels/labels.py:
```diff
@@ -1225,6 +1232,54 @@ def fill(self, coord, new_label, refresh=True):
+        try:
+            yield
+        finally:
+            self._finish_painting()
```
@Czaki I've implemented with `try: finally:` here, does this look right? I'm not sure if `_finish_painting` should be called in the `finally` block or after the `yield` in the `try`...

jni commented on napari/layers/labels/_labels_mouse_bindings.py:
```diff
@@ -26,57 +21,26 @@ def draw(layer, event):
+        if layer._mode == Mode.ERASE:
+            new_label = layer._background_label
+        else:
+            new_label = layer.selected_label
```
These lines belong outside the `with` block?

But otherwise, man, this looks so much cleaner! 😍

jni reviewed (approved):
I ❤️❤️❤️❤️ this now @DragaDoncila 😍 I only made one super minor nitpicky comment

Commit:
Remove paint event emission

Commit:
Group functions together, add data_setitem and undo_append, emit event again

DragaDoncila committed:
Merge branch 'main' into refactor-label-paint

jni committed:
Update docstrings to PEP257 & numpydoc conventions

jni committed:
Rename _undo_append

_undo_append sounds like we are undoing an append
operation. Instead, _append_to_undo_history clearly
expresses intent where it is used.

jni committed:
Docstring improvements

jni committed:
Clear redo queue instead of creating new object

tlambert03 reviewed (approved):
❤️

Czaki commented on napari/layers/labels/labels.py:
```diff
@@ -1099,11 +1140,11 @@ def _save_history(self, value):
+            self._append_to_undo_history([value])
         else:
-            self._undo_history[-1].append(value)
+            self._staged_history.append(value)
```
I cannot make suggestions becuse of deleted line so I put it in comment:

```python
if self._block_history:
    self._staged_history.append(value)
else:
    self._append_to_undo_history([value])
```

Commit:
Switch if statement to put true first
